src/plugins/data_views/server/rest_api_routes/internal/fields.ts

```diff
@@ -0,0 +1,144 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import { createHash } from 'crypto';
+import { IRouter, RequestHandler, StartServicesAccessor, KibanaRequest } from '@kbn/core/server';
+import { unwrapEtag } from '../../../common/utils';
+import { IndexPatternsFetcher } from '../../fetcher';
+import type {
+  DataViewsServerPluginStart,
+  DataViewsServerPluginStartDependencies,
+} from '../../types';
+import type { FieldDescriptorRestResponse } from '../route_types';
+import { FIELDS_PATH as path } from '../../../common/constants';
+import { parseFields, IBody, IQuery, querySchema } from './fields_for';
+import { DEFAULT_FIELD_CACHE_FRESHNESS } from '../../constants';
+
+export function calculateHash(srcBuffer: Buffer) {
+  const hash = createHash('sha1');
+  hash.update(srcBuffer);
+  return hash.digest('hex');
+}
+
+const handler: (
+  isRollupsEnabled: () => boolean,
+  getUserId: () => (kibanaRequest: KibanaRequest) => Promise<string | undefined>
+) => RequestHandler<{}, IQuery, IBody> =
+  (isRollupsEnabled, getUserIdGetter) => async (context, request, response) => {
+    const core = await context.core;
+    const uiSettings = core.uiSettings.client;
+    const { asCurrentUser } = core.elasticsearch.client;
+    const indexPatterns = new IndexPatternsFetcher(asCurrentUser, undefined, isRollupsEnabled());
+    const {
+      pattern,
+      meta_fields: metaFields,
+      type,
+      rollup_index: rollupIndex,
+      allow_no_index: allowNoIndex,
+      include_unmapped: includeUnmapped,
+    } = request.query;
+
+    let parsedFields: string[] = [];
+    let parsedMetaFields: string[] = [];
+    try {
+      parsedMetaFields = parseFields(metaFields);
+      parsedFields = parseFields(request.query.fields ?? []);
+    } catch (error) {
+      return response.badRequest();
+    }
+
+    try {
+      const { fields, indices } = await indexPatterns.getFieldsForWildcard({
+        pattern,
+        metaFields: parsedMetaFields,
+        type,
+        rollupIndex,
+        fieldCapsOptions: {
+          allow_no_indices: allowNoIndex || false,
+          includeUnmapped,
+        },
+        ...(parsedFields.length > 0 ? { fields: parsedFields } : {}),
+      });
+
+      const body: { fields: FieldDescriptorRestResponse[]; indices: string[] } = {
+        fields,
+        indices,
+      };
+
+      const etag = calculateHash(Buffer.from(JSON.stringify(body)));
```

Comment:
I don't have a particular reason to be concerned with the performance of this code but it would certainly be good to meet a higher standard than that. This gives me some confidence that this is reasonably efficient - https://blog.shevlyagin.com/2021/10/28/fastest-node-js-hashing-algorithm-for-large-strings/ - dealing with particularly large field lists could certainly impact performance but I suspect this is small fraction of the larger inefficient request. 

I can reach out to the core team for another perspective as well.

Reply:
Apparently @afharo is the person to talk to

Reply:
Am I? 😅 

I'll share my knowledge so far with perf testing: @dmlemeshko is the best person to speak to. 

When in doubt, my best recommendation is:

1. Add your API to the API scalability tests: https://github.com/elastic/kibana/tree/main/x-pack/test/scalability/apis
2. Run our API Capacity tests: https://buildkite.com/elastic/kibana-apis-capacity-testing (you can trigger it manually for `refs/pull/PR_NUMBER/head`)
3. Schedule a few runs to have better averages
4. Compare the results with `main` in https://kibana-ops-e2e-perf.kb.us-central1.gcp.cloud.es.io:9243/ (we should have APM data for both `labels.branch` that we can compare. Please, @dmlemeshko, keep me honest if this is the best place to compare them.

Regarding the hashing algo, I'd double check with @elastic/kibana-security if that's a valid one, based on the latest cert requirements.

Reply:
I'm working on creating a baseline for `fields_for_wildcard` performance over here https://github.com/elastic/kibana/pull/171956

Reply:
after speaking with @dmlemeshko I'm going to add performance journeys that use a smaller number of fields to see how scalability is affected.